File: 30.BWINF/2.runde/aufgabe2/cube.c
/* 30. Bundeswettbewerb Informatik - 2. Runde
 * Aufgabe 2 - Kool
 *
 * Autor: Matthias Rüster
 *
 * E-Mail: ruester@molgen.mpg.de
 */

#define _GNU_SOURCE

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "cube.h"

struct cube *cube_new(char *type)
{
    struct cube *c;

    if ((c = calloc(1, sizeof(*c))) == NULL) {
        perror("calloc");
        exit(EXIT_FAILURE);
    }

    c->type = convert_type(type);
    cube_init(c);

    return c;
}

struct cube *cube_new_by_type(int type)
{
    struct cube *c;

    if ((c = calloc(1, sizeof(*c))) == NULL) {
        perror("calloc");
        exit(EXIT_FAILURE);
    }

    c->type = type;
    cube_init(c);

    return c;
}

int convert_type(char *type)
{
    if (!strcasecmp("kreuzung", type))
        return KREUZUNG;

    if (!strcasecmp("geradeaus-ost-west", type)
        || !strcasecmp("geradeaus-west-ost", type))
        return GERADEAUS_OST_WEST;

    if (!strcasecmp("geradeaus-nord-sued", type)
        || !strcasecmp("geradeaus-sued-nord", type))
        return GERADEAUS_NORD_SUED;

    if (!strcasecmp("knick-nord-west", type)
        || !strcasecmp("knick-west-nord", type))
        return KNICK_NORD_WEST;

    if (!strcasecmp("knick-nord-ost", type)
        || !strcasecmp("knick-ost-nord", type))
        return KNICK_NORD_OST;

    if (!strcasecmp("knick-ost-sued", type)
        || !strcasecmp("knick-sued-ost", type))
        return KNICK_OST_SUED;

    if (!strcasecmp("knick-sued-west", type)
        || !strcasecmp("knick-west-sued", type))
        return KNICK_SUED_WEST;

    if (!strcasecmp("t-kreuzung-west", type))
        return TKREUZUNG_WEST;

    if (!strcasecmp("t-kreuzung-nord", type))
        return TKREUZUNG_NORD;

    if (!strcasecmp("t-kreuzung-ost", type))
        return TKREUZUNG_OST;

    if (!strcasecmp("t-kreuzung-sued", type))
        return TKREUZUNG_SUED;

    if (!strcasecmp("sackgasse-west", type))
        return SACKGASSE_WEST;

    if (!strcasecmp("sackgasse-nord", type))
        return SACKGASSE_NORD;

    if (!strcasecmp("sackgasse-ost", type))
        return SACKGASSE_OST;

    if (!strcasecmp("sackgasse-sued", type))
        return SACKGASSE_SUED;

    fprintf(stderr, "unbekannter Wuerfeltyp '%s'\n", type);
    exit(EXIT_FAILURE);
}

/* setze Oeffnungen anhand des Typs */
void cube_init(struct cube *c)
{
    c->north = !!(c->type & (1 << 3));
    c->east  = !!(c->type & (1 << 2));
    c->south = !!(c->type & (1 << 1));
    c->west  = !!(c->type & 1);
}

/* um 90 Grad im Uhrzeigersinn drehen; count-mal */
void cube_rotate(struct cube *c, int count)
{
    int h;

    switch (count) {
        case 1:
            h        = c->north;
            c->north = c->west;
            c->west  = c->south;
            c->south = c->east;
            c->east  = h;
        break;

        case 2:
            h        = c->north;
            c->north = c->south;
            c->south = h;
            h        = c->west;
            c->west  = c->east;
            c->east  = h;
        break;

        case 3:
            h        = c->north;
            c->north = c->east;
            c->east  = c->south;
            c->south = c->west;
            c->west  = h;
        break;

        default:
            fprintf(stderr, "bug: ungueltiges Argument count=%d\n",
                    count);
            exit(EXIT_FAILURE);
    }

    c->type = get_type(c);
}

char get_type(struct cube *c)
{
    char type;

    type = 0;

    if (c->north)
        type += (1 << 3);

    if (c->east)
        type += (1 << 2);

    if (c->south)
        type += (1 << 1);

    if (c->west)
        type += 1;

    return type;
}

char *type_to_string(struct cube *c)
{
    char *ret;

    switch (c->type) {
        case GERADEAUS_OST_WEST:
            asprintf(&ret, "Geradeaus-Ost-West");
            break;

        case GERADEAUS_NORD_SUED:
            asprintf(&ret, "Geradeaus-Nord-Sued");
            break;

        case KNICK_NORD_WEST:
            asprintf(&ret, "Knick-Nord-West");
            break;

        case KNICK_NORD_OST:
            asprintf(&ret, "Knick-Nord-Ost");
            break;

        case KNICK_OST_SUED:
            asprintf(&ret, "Knick-Ost-Sued");
            break;

        case KNICK_SUED_WEST:
            asprintf(&ret, "Knick-Sued-West");
            break;

        case KREUZUNG:
            asprintf(&ret, "Kreuzung");
            break;

        case TKREUZUNG_WEST:
            asprintf(&ret, "T-Kreuzung-West");
            break;

        case TKREUZUNG_NORD:
            asprintf(&ret, "T-Kreuzung-Nord");
            break;

        case TKREUZUNG_OST:
            asprintf(&ret, "T-Kreuzung-Ost");
            break;

        case TKREUZUNG_SUED:
            asprintf(&ret, "T-Kreuzung-Sued");
            break;

        case SACKGASSE_WEST:
            asprintf(&ret, "Sackgasse-West");
            break;

        case SACKGASSE_NORD:
            asprintf(&ret, "Sackgasse-Nord");
            break;

        case SACKGASSE_OST:
            asprintf(&ret, "Sackgasse-Ost");
            break;

        case SACKGASSE_SUED:
            asprintf(&ret, "Sackgasse-Sued");
            break;

        default:
            fprintf(stderr, "bug: unbekannter Typ %d\n", c->type);
            exit(EXIT_FAILURE);
    }

    return ret;
}

File: 30.BWINF/2.runde/aufgabe2/Cube.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;

public class Cube extends JTextArea {
    private static final long serialVersionUID = 1L;
    public static final int GERADEAUS_NORD_SUED = 0, GERADEAUS_OST_WEST = 1,
                            KNICK_NORD_OST = 2, KNICK_NORD_WEST = 3,
                            KNICK_OST_SUED = 4, KNICK_SUED_WEST = 5,
                            KREUZUNG = 6,
                            SACKGASSE_NORD = 7, SACKGASSE_OST   = 8,
                            SACKGASSE_SUED = 9, SACKGASSE_WEST  = 10,
                            TKREUZUNG_NORD = 11, TKREUZUNG_OST  = 12,
                            TKREUZUNG_SUED = 13, TKREUZUNG_WEST = 14;
    private DefaultListModel lm;
    private int type;

    public Cube(int t, int size, DefaultListModel m) {
        this.lm = m;
        this.type = t;

        setVisible(true);
        addMouseListener(new MouseListener() {
            public void mouseReleased(MouseEvent e) {}
            public void mousePressed(MouseEvent e) {}
            public void mouseExited(MouseEvent e) {}
            public void mouseEntered(MouseEvent e) {}
            public void mouseClicked(MouseEvent e) {
                if (e.getButton() == MouseEvent.BUTTON3
                    || e.getButton() == MouseEvent.BUTTON2)
                    lm.clear();

                if (e.getButton() == MouseEvent.BUTTON1) {
                    String eingabe;
                    eingabe = JOptionPane.showInputDialog("Anzahl:");

                    if (eingabe != null)
                        lm.insertElementAt(eingabe + " " + typeToString(), 0);
                }
            }
        });
    }

    private void paintBlock(Graphics g, int b, Color c) {
        g.setColor(c);
        int w = getWidth(),
            h = getHeight();

        switch(b) {
        case 1:
            g.fillRect(0, 0, w / 3, h / 3);
            break;

        case 2:
            g.fillRect(w / 3, 0, w / 3, h / 3);
            break;

        case 3:
            g.fillRect(2 * w / 3, 0, w / 3, h / 3);
            break;

        case 4:
            g.fillRect(0, h / 3, w / 3, h / 3);
            break;

        case 5:
            g.fillRect(w / 3, h / 3, w / 3, h / 3);
            break;

        case 6:
            g.fillRect(2 * w / 3, h / 3, w / 3, h / 3);
            break;

        case 7:
            g.fillRect(0, 2 * h / 3, w / 3, h / 3);
            break;

        case 8:
            g.fillRect(w / 3, 2 * h / 3, w / 3, h / 3);
            break;

        case 9:
            g.fillRect(2 * w / 3, 2 * h / 3, w / 3, h / 3);
            break;
        }
    }

    private void paintKreuzung(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 4, Color.yellow);
        paintBlock(g, 6, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private void paintGeradeausNordSued(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private void paintGeradeausOstWest(Graphics g) {
        paintBlock(g, 4, Color.yellow);
        paintBlock(g, 6, Color.yellow);
    }

    private void paintKnickNordOst(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 6, Color.yellow);
    }

    private void paintKnickNordWest(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 4, Color.yellow);
    }

    private void paintKnickOstSued(Graphics g) {
        paintBlock(g, 6, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private void paintKnickSuedWest(Graphics g) {
        paintBlock(g, 4, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private void paintTKreuzungNord(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 4, Color.yellow);
        paintBlock(g, 6, Color.yellow);
    }

    private void paintTKreuzungOst(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 6, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private void paintTKreuzungSued(Graphics g) {
        paintBlock(g, 4, Color.yellow);
        paintBlock(g, 6, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private void paintTKreuzungWest(Graphics g) {
        paintBlock(g, 2, Color.yellow);
        paintBlock(g, 4, Color.yellow);
        paintBlock(g, 8, Color.yellow);
    }

    private String typeToString() {
        switch (type) {
        case GERADEAUS_NORD_SUED:
            return "Geradeaus-Nord-Sued";

        case GERADEAUS_OST_WEST:
            return "Geradeaus-Ost-West";

        case KNICK_NORD_OST:
            return "Knick-Nord-Ost";

        case KNICK_NORD_WEST:
            return "Knick-Nord-West";

        case KNICK_OST_SUED:
            return "Knick-Ost-Sued";

        case KNICK_SUED_WEST:
            return "Knick-Sued-West";

        case KREUZUNG:
            return "Kreuzung";

        case SACKGASSE_NORD:
            return "Sackgasse-Nord";

        case SACKGASSE_OST:
            return "Sackgasse-Ost";

        case SACKGASSE_SUED:
            return "Sackgasse-Sued";

        case SACKGASSE_WEST:
            return "Sackgasse-West";

        case TKREUZUNG_NORD:
            return "T-Kreuzung-Nord";

        case TKREUZUNG_OST:
            return "T-Kreuzung-Ost";

        case TKREUZUNG_SUED:
            return "T-Kreuzung-Sued";

        case TKREUZUNG_WEST:
            return "T-Kreuzung-West";

        default:
            System.err.println("ERROR");
            System.exit(1);
        }

        return "ERROR";
    }

    public void paint(Graphics g) {
        g.setColor(Color.white);
        g.fillRect(0, 0, getWidth(), getHeight());

        for (int i = 1; i <= 9; i++)
            paintBlock(g, i, Color.gray);
        paintBlock(g, 5, Color.yellow);

        switch (type) {
        case GERADEAUS_NORD_SUED:
            paintGeradeausNordSued(g);
        break;

        case GERADEAUS_OST_WEST:
            paintGeradeausOstWest(g);
        break;

        case KNICK_NORD_OST:
            paintKnickNordOst(g);
        break;

        case KNICK_NORD_WEST:
            paintKnickNordWest(g);
        break;

        case KNICK_OST_SUED:
            paintKnickOstSued(g);
        break;

        case KNICK_SUED_WEST:
            paintKnickSuedWest(g);
        break;

        case KREUZUNG:
            paintKreuzung(g);
        break;

        case SACKGASSE_NORD:
            paintBlock(g, 2, Color.yellow);
        break;

        case SACKGASSE_OST:
            paintBlock(g, 6, Color.yellow);
        break;

        case SACKGASSE_SUED:
            paintBlock(g, 8, Color.yellow);
        break;

        case SACKGASSE_WEST:
            paintBlock(g, 4, Color.yellow);
        break;

        case TKREUZUNG_NORD:
            paintTKreuzungNord(g);
        break;

        case TKREUZUNG_OST:
            paintTKreuzungOst(g);
        break;

        case TKREUZUNG_SUED:
            paintTKreuzungSued(g);
        break;

        case TKREUZUNG_WEST:
            paintTKreuzungWest(g);
        break;

        default:
            System.err.println("ERROR");
            System.exit(1);
        }
    }
}
